--- old/src/share/vm/classfile/vmSymbols.hpp	2015-06-16 10:25:41.478505391 -0700
+++ new/src/share/vm/classfile/vmSymbols.hpp	2015-06-16 10:25:41.406507106 -0700
@@ -244,6 +244,7 @@
   template(returnType_name,                           "returnType")                               \
   template(signature_name,                            "signature")                                \
   template(slot_name,                                 "slot")                                     \
+  template(selectAlternative_name,                    "selectAlternative")                        \
                                                                                                   \
   /* Support for annotations (JDK 1.5 and above) */                                               \
                                                                                                   \
@@ -295,7 +296,8 @@
   template(setTarget_signature,                       "(Ljava/lang/invoke/MethodHandle;)V")       \
   NOT_LP64(  do_alias(intptr_signature,               int_signature)  )                           \
   LP64_ONLY( do_alias(intptr_signature,               long_signature) )                           \
-                                                                                                  \
+  template(selectAlternative_signature, "(ZLjava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodHandle;)Ljava/lang/invoke/MethodHandle;") \
+                                                                      \
   /* common method and field names */                                                             \
   template(object_initializer_name,                   "<init>")                                   \
   template(class_initializer_name,                    "<clinit>")                                 \
@@ -475,6 +477,7 @@
   template(threadgroup_string_void_signature,         "(Ljava/lang/ThreadGroup;Ljava/lang/String;)V")             \
   template(string_class_signature,                    "(Ljava/lang/String;)Ljava/lang/Class;")                    \
   template(object_object_object_signature,            "(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;") \
+  template(object_long_object_signature,              "(Ljava/lang/Object;J)Ljava/lang/Object;")                  \
   template(string_string_string_signature,            "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;") \
   template(string_string_signature,                   "(Ljava/lang/String;)Ljava/lang/String;")                   \
   template(classloader_string_long_signature,         "(Ljava/lang/ClassLoader;Ljava/lang/String;)J")             \
@@ -589,6 +592,15 @@
   template(classRedefinedCount_name,                   "classRedefinedCount")                                     \
   template(classLoader_name,                           "classLoader")                                             \
                                                                                                                   \
+  /* ObjectLayout support */                                                                                      \
+  template(org_ObjectLayout_AbstractStructuredArray,   "org/ObjectLayout/AbstractStructuredArray")                \
+  template(existenceIndicatesIntrinsic_name,           "existenceIndicatesIntrinsic")                             \
+  template(bodySize_name,                              "bodySize")                                                \
+  template(length_name,                                "length")                                                  \
+  template(elementSize_name,                           "elementSize")                                             \
+  template(paddingSize_name,                           "paddingSize")                                             \
+  template(elementClass_name,                          "elementClass")                                            \
+                                                                                                                  \
   /* trace signatures */                                                                                          \
   TRACE_TEMPLATES(template)                                                                                       \
                                                                                                                   \
@@ -866,14 +878,29 @@
    do_name(     fullFence_name,                                  "fullFence")                                           \
    do_alias(    fullFence_signature,                              void_method_signature)                                \
                                                                                                                         \
-  /* Custom branch frequencies profiling support for JSR292 */                                                          \
-  do_class(java_lang_invoke_MethodHandleImpl,               "java/lang/invoke/MethodHandleImpl")                        \
-  do_intrinsic(_profileBoolean, java_lang_invoke_MethodHandleImpl, profileBoolean_name, profileBoolean_signature,    F_S)  \
-   do_name(     profileBoolean_name,                               "profileBoolean")                                     \
-   do_signature(profileBoolean_signature,                           "(Z[I)Z")                                            \
+  /* ObjectLayout intrinsics */                                                                                         \
+   do_signature(class_long_long_long_signature,                  "(Ljava/lang/Class;JJ)J")                              \
+  do_intrinsic(_getInstanceSize,                                  sun_misc_Unsafe, getInstanceSize_name, class_long_signature, F_RN) \
+   do_name(     getInstanceSize_name,                            "getInstanceSize")                                     \
+  do_intrinsic(_getInstanceFootprintWhenContained,                sun_misc_Unsafe, getInstanceFootprintWhenContained_name, class_long_signature, F_RN) \
+   do_name(     getInstanceFootprintWhenContained_name,          "getInstanceFootprintWhenContained")                   \
+  do_intrinsic(_getContainingObjectFootprint,                     sun_misc_Unsafe, getContainingObjectFootprint_name, class_long_long_long_signature, F_RN) \
+   do_name(     getContainingObjectFootprint_name,               "getContainingObjectFootprint")                        \
+  do_intrinsic(_getContainingObjectFootprintWhenContained,        sun_misc_Unsafe, getContainingObjectFootprintWhenContained_name, class_long_long_long_signature, F_RN) \
+   do_name(     getContainingObjectFootprintWhenContained_name,  "getContainingObjectFootprintWhenContained")           \
+  do_intrinsic(_getPrePaddingInObjectFootprint,                   sun_misc_Unsafe, getPrePaddingInObjectFootprint_name, long_long_signature, F_RN) \
+   do_name(     getPrePaddingInObjectFootprint_name,             "getPrePaddingInObjectFootprint")                      \
+  do_intrinsic(_allocateHeapForElementArrayClass,                 sun_misc_Unsafe, allocateHeapForElementArrayClass_name, allocateHeapForElementArrayClass_signature, F_RN) \
+   do_name(     allocateHeapForElementArrayClass_name,           "allocateHeapForElementArrayClass")                    \
+   do_signature(allocateHeapForElementArrayClass_signature,      "(Ljava/lang/Class;[J[Ljava/lang/Class;)Ljava/lang/Object;") \
+  do_intrinsic(_constructObjectAtOffset,                          sun_misc_Unsafe, constructObjectAtOffset_name, constructObjectAtOffset_signature, F_RN) \
+   do_name(     constructObjectAtOffset_name,                    "constructObjectAtOffset")                             \
+   do_signature(constructObjectAtOffset_signature,               "(Ljava/lang/Object;JJZZJLjava/lang/reflect/Constructor;[Ljava/lang/Object;)V") \
+  do_intrinsic(_deriveContainedObjectAtOffset,                    sun_misc_Unsafe, deriveContainedObjectAtOffset_name, object_long_object_signature, F_RN) \
+   do_name(     deriveContainedObjectAtOffset_name,              "deriveContainedObjectAtOffset")                       \
                                                                                                                         \
   /* unsafe memory references (there are a lot of them...) */                                                           \
-  do_signature(getObject_signature,       "(Ljava/lang/Object;J)Ljava/lang/Object;")                                    \
+  do_alias(    getObject_signature,        object_long_object_signature)                                                \
   do_signature(putObject_signature,       "(Ljava/lang/Object;JLjava/lang/Object;)V")                                   \
   do_signature(getBoolean_signature,      "(Ljava/lang/Object;J)Z")                                                     \
   do_signature(putBoolean_signature,      "(Ljava/lang/Object;JZ)V")                                                    \
